src/core/qgsvectorlayerutils.cpp

```diff
@@ -498,6 +498,34 @@ QgsFeature QgsVectorLayerUtils::duplicateFeature( QgsVectorLayer *layer, const Q
   return newFeature;
 }
 
+std::unique_ptr<QgsVectorLayerFeatureSource> QgsVectorLayerUtils::getFeatureSource( QWeakPointer<QgsVectorLayer> layer )
+{
+  std::unique_ptr<QgsVectorLayerFeatureSource> featureSource;
+
+  auto getFeatureSource = [ layer, &featureSource ]
+  {
+    QgsVectorLayer *lyr = layer.data();
```

Comment:
Could you add an assert in here (for qt 10 and above) checking that it's the main thread?

And then with a simple unit test checking that a source is returned we should be protected here.